### PR TITLE
demo native marketplace

### PR DESCRIPTION
**Pull request by tbro**
Adds a second demo-native workflow specifically for marketplace. 

Note this is a ~~quick~~ hack, in followups we should:

  * remove duplicated files (we can have more robust configuration)
  * address #1888 
  * remove it from PR. We can run it from [here](https://github.com/EspressoSystems/espresso-sequencer/actions/workflows/test-demo-native-marketplace.yml) once it is merged.

### CODE REVIEW

**Review by elliedavidson — COMMENTED** [review]
LGTM!

File: data/genesis/demo-marketplace.toml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+chain_id = 999999999
+base_fee = '0 wei'
+max_block_size = '1mb'
+fee_recipient = '0x0000000000000000000000000000000000000000'
+fee_contract = '0xa15bb66138824a1c7167f5e85b957d04dd34e468'
+
+[header]
+timestamp = "1970-01-01T00:00:00Z"
+
+[[upgrade]]
```
Comment (imabdulbasit) [inline data/genesis/demo-marketplace.toml:17]: We can remove the fee upgrade from this toml file

Comment (tbro) [reply data/genesis/demo-marketplace.toml:17]: Is `timestamp` part of the upgrade?

Comment (imabdulbasit) [reply data/genesis/demo-marketplace.toml:17]: No, I meant to comment on the upgrade L17

File: data/genesis/demo-marketplace.toml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+version = "0.2"
+start_proposing_view = 5
+stop_proposing_view = 15
+
+[upgrade.fee]
```
Comment (imabdulbasit) [inline data/genesis/demo-marketplace.toml:22]: L17 to L29 should be removed

Comment (tbro) [reply data/genesis/demo-marketplace.toml]: f8955d1f5b286330339dea212f03926e48461c5b

**Comment by sveitser** [comment]
Looking at the diff of the two process compose files it seems that most of the changes can be done by setting some env vars and we add one service.

```diff
diff process-compose.yaml process-compose-mp.yml                                                 0.274s  tb/marketplace/demo-native-mp  ❄  15:32:26
8,9c8,9
<   - ESPRESSO_SEQUENCER_GENESIS_FILE=data/genesis/demo.toml
<   - ESPRESSO_BUILDER_GENESIS_FILE=data/genesis/demo.toml
---
>   - ESPRESSO_SEQUENCER_GENESIS_FILE=data/genesis/demo-marketplace.toml
>   - ESPRESSO_BUILDER_GENESIS_FILE=data/genesis/demo-marketplace.toml
370,371c370
<
<   # We use KeyDB (a Redis variant) to maintain consistency between
---
>   # We use KeyDB (a Redis variant) to maintain consistency between
459a459,460
>       - ESPRESSO_SUBMIT_TRANSACTIONS_MIN_NAMESPACE=1
>       - ESPRESSO_SUBMIT_TRANSACTIONS_MAX_NAMESPACE=6
479a481,502
>       - ESPRESSO_SUBMIT_TRANSACTIONS_MIN_NAMESPACE=1
>       - ESPRESSO_SUBMIT_TRANSACTIONS_MAX_NAMESPACE=3
>     depends_on:
>       sequencer0:
>         condition: process_healthy
>       permissionless-builder:
>         condition: process_healthy
>     readiness_probe:
>       http_get:
>         scheme: http
>         host: localhost
>         port: $ESPRESSO_SUBMIT_TRANSACTIONS_PRIVATE_PORT
>         path: /healthcheck
>       failure_threshold: 100
>
>   submit-transactions-private-fallback:
>     command: submit-transactions
>     environment:
>       - ESPRESSO_SUBMIT_TRANSACTIONS_PORT=$ESPRESSO_SUBMIT_TRANSACTIONS_PRIVATE_PORT
>       - ESPRESSO_SUBMIT_TRANSACTIONS_SUBMIT_URL=http://localhost:$ESPRESSO_FALLBACK_BUILDER_SERVER_PORT/txn_submit
>       - ESPRESSO_SUBMIT_TRANSACTIONS_MIN_NAMESPACE=4
>       - ESPRESSO_SUBMIT_TRANSACTIONS_MAX_NAMESPACE=6
494c517
<     disabled: true
---
>     # disabled: true
501d523
<       - ESPRESSO_MARKETPLACE_BUILDER_NAMESPACE=10000
504a527
>       - ESPRESSO_MARKETPLACE_BUILDER_NAMESPACE=1,2,3
519c542
<     disabled: true
---
>     # disabled: true
527d549
<       - ESPRESSO_MARKETPLACE_BUILDER_NAMESPACE=10000
544a567
>     disabled: true
667,668c690
<     command:
<       dev-rollup register --ns 1; dev-rollup register --ns 2; dev-rollup register --ns 3
---
>     command: dev-rollup register --ns 1; dev-rollup register --ns 2; dev-rollup register --ns 3
670a693
>       - ESPRESSO_MARKETPLACE_RESERVE_BUILDER_URL=http://localhost:$ESPRESSO_BUILDER_SERVER_PORT
```

Was wondering if you considered adding the difference as an override file: https://f1bonacc1.github.io/process-compose/configuration/#merge-2-or-more-configuration-files-with-override-values

**Comment by tbro** [comment]
> Was wondering if you considered adding the difference as an override file

I wasn't aware I could do that.

File: .github/workflows/test-demo-native-marketplace.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+      - name: Enable Rust Caching
+        uses: Swatinem/rust-cache@v2
+
+      - name: Build
+        run: cargo build --locked --release
+
+      - name: Test Demo Marketplace
+        run: |
+          export PATH="$PWD/target/release:$PATH"
```
Comment (sveitser) [inline .github/workflows/test-demo-native-marketplace.yml:51]: Add `set -o pipefail` like for the other smoke test?

File: process-compose-mp.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MIN_NAMESPACE=1
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MAX_NAMESPACE=6
+
+  submit-transactions-private:
+    command: submit-transactions
+    environment:
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MIN_NAMESPACE=1
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MAX_NAMESPACE=3
+
+  submit-transactions-private-fallback:
```
Comment (imabdulbasit) [inline process-compose-mp.yml:19]: I am thinking that if we also need another load generator for reserve builder? so we the submit-transactions-private load generator is for legacy one

Comment (tbro) [reply process-compose-mp.yml:19]: I think something like 8e65d4a492e0930336bbce15264cdcfc33403a86. I'm not sure of the namespaces though.

Comment (tbro) [reply process-compose-mp.yml:19]: all the private load generators are currently failing, so looking into it. Maybe there's a wrong port somewhere: 

```
│2024-09-03T20:33:25.558719Z ERROR submit_transactions: failed to submit     ││
││batch of 3 transactions err=Custom { message: "error sending request for    ││
││url (http://localhost:31003/txn_submit/batch): client error (Connect)",     ││
││status: StatusCode(500) }  
```

File: process-compose-mp.yml [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    command: submit-transactions
+    environment:
+      - ESPRESSO_SUBMIT_TRANSACTIONS_PORT=$ESPRESSO_SUBMIT_TRANSACTIONS_PRIVATE_PORT
+      - ESPRESSO_SUBMIT_TRANSACTIONS_SUBMIT_URL=http://localhost:$ESPRESSO_FALLBACK_BUILDER_SERVER_PORT/txn_submit
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MIN_NAMESPACE=4
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MAX_NAMESPACE=6
+    depends_on:
+      sequencer0:
+        condition: process_healthy
+      permissionless-builder:
```
Comment (imabdulbasit) [inline process-compose-mp.yml:29]: depends on fallback builder?

File: process-compose-mp.yml [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    command: submit-transactions
+    environment:
+      - ESPRESSO_SUBMIT_TRANSACTIONS_PORT=$ESPRESSO_SUBMIT_TRANSACTIONS_PRIVATE_PORT
+      - ESPRESSO_SUBMIT_TRANSACTIONS_SUBMIT_URL=http://localhost:$ESPRESSO_BUILDER_SERVER_PORT/txn_submit
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MIN_NAMESPACE=4
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MAX_NAMESPACE=6
+    depends_on:
+      sequencer0:
+        condition: process_healthy
+      permissionless-builder:
```
Comment (imabdulbasit) [inline process-compose-mp.yml:49]: same here

File: process-compose-mp.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        host: localhost
+        port: $ESPRESSO_SUBMIT_TRANSACTIONS_PRIVATE_PORT
+        path: /healthcheck
+      failure_threshold: 100
+
+  submit-transactions-private-reserve:
+    command: submit-transactions
+    environment:
+      - ESPRESSO_SUBMIT_TRANSACTIONS_PORT=$ESPRESSO_SUBMIT_TRANSACTIONS_PRIVATE_PORT
+      - ESPRESSO_SUBMIT_TRANSACTIONS_SUBMIT_URL=http://localhost:$ESPRESSO_BUILDER_SERVER_PORT/txn_submit
```
Comment (imabdulbasit) [inline process-compose-mp.yml:43]: different port for reserve builder

**Review by imabdulbasit — DISMISSED** [review]

File: process-compose-mp.yml (hunk 7 above)
Comment (tbro) [reply process-compose-mp.yml:43]: marketplace-reserve-builder has port set to `$ESPRESSO_BUILDER_SERVER_PORT`

Comment (imabdulbasit) [reply process-compose-mp.yml:43]: I thought we'd be running all three builders at once in marketplace native demo but I guess its fine to use legacy port for reserve one

File: process-compose-mp.yml [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  - ESPRESSO_SEQUENCER_GENESIS_FILE=data/genesis/demo-marketplace.toml
+  - ESPRESSO_BUILDER_GENESIS_FILE=data/genesis/demo-marketplace.toml
+processes:
+
+  marketplace-reserve-builder:
+    command: marketplace-builder
+    ports:
+      - "$ESPRESSO_RESERVE_BUILDER_SERVER_PORT:$ESPRESSO_RESERVE_BUILDER_SERVER_PORT"
+    environment:
+      - ESPRESSO_RESERVE_BUILDER_SERVER_PORT=$ESPRESSO_RESERVE_BUILDER_SERVER_PORT
```
Comment (imabdulbasit) [inline process-compose-mp.yml:13]:
```suggestion
      - ESPRESSO_BUILDER_SERVER_PORT=$ESPRESSO_RESERVE_BUILDER_SERVER_PORT
```

File: process-compose-mp.yml [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    disabled: true
+
+  submit-transactions-public:
+    command: submit-transactions
+    environment:
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MIN_NAMESPACE=1
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MAX_NAMESPACE=6
+
+  submit-transactions-private:
+    disabled: true
```
Comment (sveitser) [inline process-compose-mp.yml:68]: @tbro Not sure if that's the only issue but if we don't run anything on the `ESPRESSO_SUBMIT_TRANSACTIONS_PRIVATE_PORT` we need to check the health check

https://github.com/EspressoSystems/espresso-sequencer/blob/f5e7be1e49e3492e6c462f367878b07a9670d645/scripts/smoke-test-demo#L17

https://github.com/EspressoSystems/espresso-sequencer/blob/f5e7be1e49e3492e6c462f367878b07a9670d645/scripts/smoke-test-demo#L72

File: process-compose-mp.yml [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+  submit-transactions-private-fallback:
+    command: submit-transactions
+    environment:
+      - ESPRESSO_SUBMIT_TRANSACTIONS_PORT=$ESPRESSO_SUBMIT_TRANSACTIONS_PRIVATE_FALLBACK_PORT
+      - ESPRESSO_SUBMIT_TRANSACTIONS_SUBMIT_URL=http://localhost:$ESPRESSO_FALLBACK_BUILDER_SERVER_PORT/txn_submit
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MIN_NAMESPACE=4
+      - ESPRESSO_SUBMIT_TRANSACTIONS_MAX_NAMESPACE=6
```
Comment (shenkeyao) [inline process-compose-mp.yml:76]: Looks like we set `ESPRESSO_SUBMIT_TRANSACTIONS_MIN_NAMESPACE` and `ESPRESSO_SUBMIT_TRANSACTIONS_MAX_NAMESPACE` to 1 and 3 for the reserve builder, and 4 and 6 for the fallback builder, but for either builder, `ESPRESSO_MARKETPLACE_BUILDER_NAMESPACE` is 10000. Can we change `ESPRESSO_MARKETPLACE_BUILDER_NAMESPACE` to 2 or maybe adjust the namespace range?

Comment (tbro) [reply process-compose-mp.yml:76]: b37f7b284674d6207c6d5974e0124ab68f12b65d

**Review by shenkeyao — APPROVED** [review]

**Pull request merged**